### PR TITLE
[M2M] Fixed comments

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: src/ralph/admin/static/auto-complete-helpers.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
+function getCallerNode(win) {
+    var name = windowname_to_id(win.name);
+    var elem = document.querySelector('#' + name);
+    return elem;
+}
+
+function dismissRelatedLookupPopup(win, chosenId) {
+    debugger;
```
Comment (mkurek) [inline src/ralph/admin/static/auto-complete-helpers.js]: remove it. and every other debugger here :)

File: src/ralph/admin/static/auto-complete-helpers.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    if (elem.nodeName.toUpperCase() === 'AUTO-COMPLETE') {
+        elem.updateById(newId);
+    } else if (elem) {
+        var elemName = elem.nodeName.toUpperCase();
+        if (elemName == 'SELECT') {
+            o = new Option(newRepr, newId);
+            elem.options[elem.options.length] = o;
+            o.selected = true;
+        } else if (elemName == 'INPUT') {
+            if (elem.className.indexOf('vManyToManyRawIdAdminField') != -1 && elem.value) {
```
Comment (mkurek) [inline src/ralph/admin/static/auto-complete-helpers.js:43]: this code seems very similar to the one in `dismissRelatedLookupPopup` - could you extract it to one function?

File: src/ralph/admin/widgets.py [hunk 3]
```diff
@@ -122,10 +127,11 @@ def render_options(self, choices, selected_choices):
 
 
 class AutocompleteWidget(forms.TextInput):
+    separator = ','
```
Comment (mkurek) [inline src/ralph/admin/widgets.py]: maybe `multivalue_separator`?

File: src/ralph/admin/widgets.py [hunk 4]
```diff
@@ -156,14 +161,63 @@ def get_related_url(self, info, action, *args):
...
+        polymorphic_descendants = getattr(
+            self.rel_to, '_polymorphic_descendants', []
+        )
+
+        if polymorphic_descendants:
+            # Check if model after which we are looking for is polymorphic
+            # if they are also looking for the models of its dependencies
+            # or by limit_models defined in model
+            polymorphic_models = polymorphic_descendants
+            limit_models = getattr(self.field, 'limit_models', [])
```
Comment (mkurek) [inline src/ralph/admin/widgets.py:198]: could you add some comment, that usually all polymorphic descendants are processed here, but they could be limited only to particular models using `limit_models` attribute of field?

File: src/ralph/admin/widgets.py [hunk 5]
```diff
@@ -156,14 +161,63 @@ def get_related_url(self, info, action, *args):
...
-            print('name', name)
+        if self.multi:
             if value:
-                value = value.split(',')
+                value = value.split(self.separator)
             else:
                 value = []
         return value
 
+    def get_search_fields(self):
```
Comment (mkurek) [inline src/ralph/admin/widgets.py:188]: we should think about caching value of this function

File: src/ralph/admin/widgets.py [hunk 6]
```diff
@@ -172,14 +226,13 @@ def render(self, name, value, attrs=None):
...
-            value = ','.join(force_text(v) for v in value)
+            attrs['multi'] = 'true'
+            value = self.separator.join(force_text(v) for v in value)
         else:
             value = value or ""
 
-        searched_fields = self.admin_site._registry[self.rel_to].search_fields
-        searched_fields_text = ', '.join(sorted(searched_fields))
-        searched_fields_info = "Search by: {}".format(searched_fields_text)
+        searched_fields = self.get_search_fields()
```
Comment (mkurek) [inline src/ralph/admin/widgets.py]: `searched_fields` -> `search_fields`

File: src/ralph/admin/widgets.py [hunk 7]
```diff
@@ -172,14 +226,13 @@ def render(self, name, value, attrs=None):
...
+        searched_fields_info = self.render_search_fields_info(dict(searched_fields))
```
Comment (mkurek) [inline src/ralph/admin/widgets.py]: ditto

File: src/ralph/admin/widgets.py [hunk 8]
```diff
@@ -156,14 +161,63 @@ def get_related_url(self, info, action, *args):
...
+                        related_model, field_name
+                    )
+                    key = str(related_model._meta.verbose_name)
+                    searched_fields_tooltip[key].append(
+                        str(field.verbose_name)
+                    )
+        else:
+            for field_name in self.admin_site._registry[self.rel_to].search_fields:
+                try:
+                    field = get_field_by_relation_path(
```
Comment (mkurek) [inline src/ralph/admin/widgets.py]: this block is the same as above - you could extract it to some function and call it here and above (to process search fields of single model)

File: src/ralph/admin/static/auto-complete-helpers.js (hunk 1 above)
Comment (ghost) [inline src/ralph/admin/static/auto-complete-helpers.js]: done

File: src/ralph/admin/widgets.py (hunk 3 above)
Comment (ghost) [inline src/ralph/admin/widgets.py]: done

File: src/ralph/admin/widgets.py (hunk 4 above)
Comment (ghost) [inline src/ralph/admin/widgets.py:198]: I think such explanations should be included here https://github.com/allegro/ralph/blob/ng/src/ralph/lib/mixins/fields.py#L69 (where feature is defined) rather then every its use.
Can I add your comment there?

File: src/ralph/admin/widgets.py (hunk 6 above)
Comment (ghost) [inline src/ralph/admin/widgets.py]: done

File: src/ralph/admin/widgets.py (hunk 7 above)
Comment (ghost) [inline src/ralph/admin/widgets.py]: done

File: src/ralph/admin/widgets.py (hunk 4 above)
Comment (ghost) [inline src/ralph/admin/widgets.py:198]: done

File: src/ralph/admin/widgets.py (hunk 8 above)
Comment (ghost) [inline src/ralph/admin/widgets.py]: done

File: src/ralph/admin/static/auto-complete-helpers.js (hunk 2 above)
Comment (ghost) [inline src/ralph/admin/static/auto-complete-helpers.js:43]: I can't see the pattern, could you be more precise? :)

File: src/ralph/admin/widgets.py (hunk 5 above)
Comment (ghost) [inline src/ralph/admin/widgets.py:188]: Should I decorate it with `@cached_property`?

**Comment by szok** [comment]
:ok:

**Comment by mkurek** [comment]
:ok:

**Pull request merged**